File: server/src/collectors/openmeteoJma.js
/**
 * Open-Meteo JMA mirror - free, anonymous
 * https://api.open-meteo.com/v1/jma
 */

import { fetchJson } from './_liveHelpers.js';

const CITIES = [
  { name: 'Sapporo', lat: 43.06, lon: 141.35 },
  { name: 'Sendai', lat: 38.27, lon: 140.87 },
  { name: 'Tokyo', lat: 35.69, lon: 139.69 },
  { name: 'Yokohama', lat: 35.45, lon: 139.64 },
  { name: 'Nagoya', lat: 35.18, lon: 136.91 },
  { name: 'Osaka', lat: 34.69, lon: 135.50 },
  { name: 'Hiroshima', lat: 34.40, lon: 132.46 },
  { name: 'Fukuoka', lat: 33.59, lon: 130.40 },
  { name: 'Naha', lat: 26.21, lon: 127.68 },
];
const TIMEOUT_MS = 8000;

async function fetchCity(c) {
  const url = `https://api.open-meteo.com/v1/jma?latitude=${c.lat}&longitude=${c.lon}&current=temperature_2m,wind_speed_10m,precipitation`;
  const d = await fetchJson(url, { timeoutMs: TIMEOUT_MS });
  return d?.current ?? null;
}

export default async function collectOpenmeteoJma() {
  const out = await Promise.all(CITIES.map(fetchCity));
  let source = 'live';
  const features = [];
  for (let i = 0; i < CITIES.length; i++) {
    const c = CITIES[i];
    const cur = out[i];
    if (cur) {
      features.push({
        type: 'Feature',
        geometry: { type: 'Point', coordinates: [c.lon, c.lat] },
        properties: {
          city: c.name,
          temperature_c: cur.temperature_2m ?? null,
          wind_mps: cur.wind_speed_10m ?? null,
          precipitation_mm: cur.precipitation ?? null,
          observed_at: cur.time ?? null,
          source: 'open_meteo_jma',
        },
      });
    }
  }
  if (features.length === 0) {
    source = 'seed';
    for (const c of CITIES) {
      features.push({
        type: 'Feature',
        geometry: { type: 'Point', coordinates: [c.lon, c.lat] },
        properties: { city: c.name, temperature_c: 15, wind_mps: 3, precipitation_mm: 0, source: 'open_meteo_seed' },
      });
    }
  }
  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source,
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      description: 'Open-Meteo JMA free weather mirror',
    },
  };
}

File: native/collectors/environment/sources/openmeteo_jma.c
/* collectors/environment/sources/openmeteo_jma.c
 * Port of server/src/collectors/openmeteoJma.js.
 * Open-Meteo JMA mirror — one current-weather GET per fixed city, then a
 * FeatureCollection. Keyless. SEED branch (features.length===0) NOT ported. */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../lib/geojson.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TIMEOUT_MS 8000

static const struct { const char *name; double lat, lon; } CITIES[] = {
  {"Sapporo",   43.06, 141.35},
  {"Sendai",    38.27, 140.87},
  {"Tokyo",     35.69, 139.69},
  {"Yokohama",  35.45, 139.64},
  {"Nagoya",    35.18, 136.91},
  {"Osaka",     34.69, 135.50},
  {"Hiroshima", 34.40, 132.46},
  {"Fukuoka",   33.59, 130.40},
  {"Naha",      26.21, 127.68},
};
#define NCITY ((int)(sizeof(CITIES) / sizeof(CITIES[0])))

static void add_or_null(cJSON *p, const char *k, cJSON *v) {
  cJSON_AddItemToObject(p, k, v ? cJSON_Duplicate(v, 1) : cJSON_CreateNull());
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  cJSON *features = cJSON_CreateArray();
  for (int i = 0; i < NCITY; i++) {
    char url[256];
    snprintf(url, sizeof url,
      "https://api.open-meteo.com/v1/jma?latitude=%g&longitude=%g"
      "&current=temperature_2m,wind_speed_10m,precipitation",
      CITIES[i].lat, CITIES[i].lon);
    cJSON *d = feed_get_json(ctx->http, url, TIMEOUT_MS);
    cJSON *cur = d ? cJSON_GetObjectItem(d, "current") : NULL;
    if (cur && !cJSON_IsNull(cur)) {
      cJSON *f = cJSON_CreateObject();
      cJSON_AddStringToObject(f, "type", "Feature");
      cJSON *g = cJSON_CreateObject();
      cJSON_AddStringToObject(g, "type", "Point");
      cJSON *co = cJSON_CreateArray();
      cJSON_AddItemToArray(co, cJSON_CreateNumber(CITIES[i].lon));
      cJSON_AddItemToArray(co, cJSON_CreateNumber(CITIES[i].lat));
      cJSON_AddItemToObject(g, "coordinates", co);
      cJSON_AddItemToObject(f, "geometry", g);

      cJSON *p = cJSON_CreateObject();           /* EXACT JS key order */
      cJSON_AddStringToObject(p, "city", CITIES[i].name);
      add_or_null(p, "temperature_c",    cJSON_GetObjectItem(cur, "temperature_2m"));
      add_or_null(p, "wind_mps",         cJSON_GetObjectItem(cur, "wind_speed_10m"));
      add_or_null(p, "precipitation_mm", cJSON_GetObjectItem(cur, "precipitation"));
      add_or_null(p, "observed_at",      cJSON_GetObjectItem(cur, "time"));
      cJSON_AddStringToObject(p, "source", "open_meteo_jma");
      cJSON_AddItemToObject(f, "properties", p);
      cJSON_AddItemToArray(features, f);
    }
    if (d) cJSON_Delete(d);
  }

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[openmeteo-jma] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def openmeteo_jma_def = {
  .id = "openmeteo-jma", .collector = "environment",
  .name = "Open-Meteo JMA mirror", .name_ja = "Open-Meteo JMA",
   .update_interval_sec = 3600, .run = run,
};
REGISTER_SOURCE(openmeteo_jma_def)
